host/src/FunctionsNetHost/Grpc/GrpcClient.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Threading.Channels;
+using Google.Protobuf;
+using Grpc.Core;
+using Grpc.Net.Client;
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+using static Microsoft.Azure.Functions.Worker.Grpc.Messages.FunctionRpc;
+
+namespace FunctionsNetHost.Grpc
+{
+    internal sealed class GrpcClient
+    {
+        private readonly Channel<StreamingMessage> _outgoingMessageChannel;
+        private readonly IncomingGrpcMessageHandler _messageHandler;
+        private readonly GrpcWorkerStartupOptions _grpcWorkerStartupOptions;
+
+        internal GrpcClient(GrpcWorkerStartupOptions grpcWorkerStartupOptions, AppLoader appLoader)
+        {
+            _grpcWorkerStartupOptions = grpcWorkerStartupOptions;
+            var channelOptions = new UnboundedChannelOptions
+            {
+                SingleWriter = false,
+                SingleReader = false,
+                AllowSynchronousContinuations = true
+            };
+
+            _outgoingMessageChannel = Channel.CreateUnbounded<StreamingMessage>(channelOptions);
+
+            _messageHandler = new IncomingGrpcMessageHandler(appLoader);
+        }
+
+        internal async Task InitAsync()
+        {
+            var endpoint = $"http://{_grpcWorkerStartupOptions.Host}:{_grpcWorkerStartupOptions.Port}";
+            Logger.LogTrace($"Grpc service endpoint:{endpoint}");
+
+            var functionRpcClient = CreateFunctionRpcClient(endpoint);
+            var eventStream = functionRpcClient.EventStream();
+
+            await SendStartStreamMessageAsync(eventStream.RequestStream);
+
+            var readerTask = StartReaderAsync(eventStream.ResponseStream);
+            var writerTask = StartWriterAsync(eventStream.RequestStream);
+            _ = StartInboundMessageForwarding();
+            _ = StartOutboundMessageForwarding();
+
+            await Task.WhenAll(readerTask, writerTask);
```

Comment:
Do we want to await in `InitAsync`? This won't complete while processing messages.

Reply:
Yes, If we do not await `InitAsync`, we will start the task for InitAsync and just go to next line (which is exiting main)